Fix bug #171: support command execute in linux or windows

ntwrick opened the pull request:
Tested in linux and win10.

Thanks to @haifenghuang.

ntwrick commented:
We have evaluator tests that run `posix commands` such as `echo`, et al. that do not work in windows.
Suppressing these for now. Need to find a way to run windows equivalent tests. Not sure what is the best way to do this.

Meanwhile, windows commands do work so we should merge this PR and make a new issue for tracking  automated testing for windows.

odino commented on evaluator/evaluator_test.go:
```diff
@@ -508,7 +509,10 @@ b, c = [1, 2]; b`, 1},
 				`, 10},
 	}
-
+	if runtime.GOOS == "windows" {
```
I'd be worried not testing all of this stuff on windows. Maybe we can do the opposite -- check if we're not running on windows and push the test case inside the `tests`?

odino commented on evaluator/evaluator_test.go:
```diff
@@ -1022,7 +1026,10 @@ func TestBuiltinProperties(t *testing.T) {
 		{`[].a`, "invalid property 'a' on type ARRAY"},
 	}
-
+	if runtime.GOOS == "windows" {
```
same here

odino commented on evaluator/evaluator_test.go:
```diff
@@ -1101,7 +1108,10 @@ func TestCommand(t *testing.T) {
 		{"`echo \\$CONTEXT`", "abs"},
 	}
-
+	if runtime.GOOS == "windows" {
```
here it's different instead. Since the majority of these are unix commands I think it's ok to keep these out of the windows test suite for now, so :+1: I just have to remember that when we merge this I should open an issue to make sure we have tests for win commands as well

ntwrick commented:
@odino: 

I created a window-specific commands test and removed bash-isms from other tests replacing them with builtins. Also, in windows REPL `$(echo string) == "string\r"`, so the tests use this because you can't get rid of the `CRLF` from windows `echo`. Also, there is no equivalent of `xargs` in windows, so the pipe tests have been removed in windows.

However, in the travis windows build, the commands tests are failing with nothing but a `' '` to show for it. Any ideas?

odino commented on evaluator/evaluator_test.go:
```diff
@@ -1020,16 +1016,12 @@ func TestBuiltinProperties(t *testing.T) {
 		{`"a".ok`, false},
 		{`"a".inv`, "invalid property 'inv' on type STRING"},
-		{"a = $(date);\na.ok", true},
+		{"a = cd();\na.ok", true},
```
we should keep a test that checks for `cmd.ok` as `cd` is a special case on its own

odino commented:
@ntwrick could [this](https://github.com/abs-lang/abs/blob/master/evaluator/evaluator.go#L1111) be of help? I remove extra newlines on linux, so it could be we should do the same on win as well?

odino commented:
This is what happens when I remove the code for trimming altogether, should give you a bit more clarity:

https://travis-ci.com/abs-lang/abs/jobs/177535279

odino commented:
@ntwrick BTW I'd like to reduce the scope of `1.2.0` to the current open PRs. Once they're reviewed and merged I think it's time for a release :) 1.2 will by then have:

* win commands
* background commands
* fixes to .json
* abs history

It would be great if we could also include the `include(file)` stuff you've been working on -- up to you if you're ready to finalize that branch as well.

ntwrick commented:
Using `strings.TrimSpace()` at return from eval command. This works in both linux and windows. Removed `\r` in windows command tests. Also removed `$(echo \$CONTEXT)` as OS env $VARs are not defined in windows.

ntwrick commented:
@odino: Working on your suggestions to #163 in branch `absrc-source-file`. Nearly complete but may need another day to finish.

odino commented:
Yo-hoo! Amazing!

On Thu, Feb 14, 2019 at 12:10 AM Erich (Rick) Arnold <
notifications@github.com> wrote:

> Using strings.TrimSpace() at return from eval command. This works in both
> linux and windows. Removed \r in windows command tests. Also removed $(echo
> \$CONTEXT) as OS env $VARs are not defined in windows.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/abs-lang/abs/pull/180#issuecomment-463349666>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUC5IpzFjSA_tDZv_JRUylRbFyOfq2Nks5vNHFOgaJpZM4a38bL>
> .
>


-- 
Alex Nadalin
http://www.linkedin.com/in/alessandronadalin
www.odino.org
www.twitter.com/_odino_

odino commented:
On Thu, Feb 14, 2019 at 12:15 AM Erich (Rick) Arnold <
notifications@github.com> wrote:

> @odino <https://github.com/odino>: Working on your suggestions to #163
> <https://github.com/abs-lang/abs/issues/163> in branch absrc-source-file.
> Nearly complete but may need another day to finish.
>

Thanks a bunch, take as long as you need! I'm happy to hold off even for a
week if you'd like. You've been helping so much and I don't want to
pressurize you :)


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/abs-lang/abs/pull/180#issuecomment-463351154>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUC5FnOvDly5Go-zAO9egmgzmAch4iKks5vNHJRgaJpZM4a38bL>
> .
>


-- 
Alex Nadalin
http://www.linkedin.com/in/alessandronadalin
www.odino.org
www.twitter.com/_odino_
